### PR TITLE
Add aud (audience) to JWT properties

### PR DESCRIPTION
**Pull request by BaurzhanSakhariev**
Follow up to https://github.com/crate/crate/pull/15650

Summary:
1. CREATE/ALTER user support aud property.
2. jwt authentication compares token.aud with user's aud if exists or cluster id.
3. Uniqueness check is done based on iss/username, see changes in `RolesMetadata.contains`
4. `JWT_TOKEN` constant is regenerated with "aud" as it's used in happy path scenarios or non-happy path scenarios where `aud` shouldn't  be a problem

### CODE REVIEW

**Comment by BaurzhanSakhariev** [comment]
https://jenkins.crate.io/job/CrateDB/job/crate_on_pr/13451/testReport/junit/io.crate.integrationtests.disruption.seqno/ConcurrentSeqNoVersioningIT/testSeqNoCASLinearizability/

retest this please

File: docs/admin/auth/methods.rst [hunk 1]
```diff
@@ -123,12 +123,20 @@ CrateDB will validate the token and match it to a user created with ``CREATE
...
+
+``aud`` - `aud`_.
+
 ``username`` - user name in a third party app.
 
 ``iss`` and  ``username`` values must match the values created by
-``CREATE USER`` statement. See :ref:`create-user-jwt` for details.
+``CREATE USER`` statement. If ``aud`` is provided in ``CREATE USER`` statement,
+token's ``aud`` value must match value created by ``CREATE USER`` statement.
+Otherwise, token's ``aud`` value must match cluster id.
```
Comment (seut) [inline docs/admin/auth/methods.rst:138]:
```suggestion
``CREATE USER`` statement. If ``aud`` has not been defined on the ``CREATE USER`` statement,
the cluster id is used and must match the token's ``aud`` value.
```

File: docs/sql/statements/alter-role.rst [hunk 2]
```diff
@@ -56,15 +56,20 @@ are supported to alter an existing user account:
...
+     ALTER USER john WITH (jwt = {"iss" = 'new_issuer', "username" = 'john.smith', "aud" = 'new_aud'})
 
   New JWT properties must not coincide with JWT properties of another user.
 
   ``NULL`` removes the JWT properties from the user.
 
+.. NOTE::
+
+   ``jwt = {...}`` overrides existing jwt properties. If an optional property
+   is not provided, an existing value will be replaced by ``NULL``.
```
Comment (seut) [inline docs/sql/statements/alter-role.rst:71]:
```suggestion
   ``jwt = {...}`` overrides existing jwt properties. If an optional property
   is not provided, an existing value will be discarded.
```

File: docs/sql/statements/create-user.rst [hunk 3]
```diff
@@ -85,13 +85,15 @@ The following ``user_parameter`` are supported to define a new user account:
...
-     CREATE USER john WITH (jwt = {"iss" = 'https://example.com', "username" = 'test@example.com'})
+     CREATE USER john WITH (jwt = {"iss" = 'https://example.com', "username" = 'test@example.com', "aud" = 'test_aud'})
 
-  `iss`_ is a JWK endpoint, containing public keys.
+  `iss`_ is a JWK endpoint, containing public keys. Required field.
 
-  ``username`` is a user name in a third party app.
+  ``username`` is a user name in a third party app. Required field.
+
+  `aud`_ is a recipient that the JWT is intended for. Optional field.
```
Comment (seut) [inline docs/sql/statements/create-user.rst:96]:
```suggestion
  `aud`_ is a recipient that the JWT is intended for. Optional field. If not provided, the cluster id is used (default).
```

File: server/src/main/java/io/crate/auth/HostBasedAuthentication.java [hunk 4]
```diff
@@ -96,11 +96,17 @@ static SSL parseValue(String value) {
     private final Roles roles;
     private final DnsResolver dnsResolver;
 
+    private final Supplier<String> clusterId;
+
     @Inject
```
Comment (seut) [inline server/src/main/java/io/crate/auth/HostBasedAuthentication.java:101]: Not related to this PR, but I think we instantiate it ourselves nowadays so the guide annotations could be removed?

File: server/src/test/java/io/crate/auth/UserAuthenticationMethodTest.java [hunk 5]
```diff
@@ -145,25 +148,90 @@ public void test_jwt_authentication() throws Exception {
...
+        );
+        JWTAuthenticationMethod jwtAuth = new JWTAuthenticationMethod(
+            roles,
+            jwkProviderFunction(null),
+            () -> clusterId
+        );
+        assertThat(jwtAuth.name()).isEqualTo("jwt");
+
+        Credentials credentials = new Credentials(JWT_TOKEN);
+        assertThat(credentials.username()).isNull();
```
Comment (seut) [inline server/src/test/java/io/crate/auth/UserAuthenticationMethodTest.java:173]: Same here, relevant for the test?
```suggestion
```

File: server/src/test/java/io/crate/auth/UserAuthenticationMethodTest.java [hunk 6]
```diff
@@ -145,25 +148,90 @@ public void test_jwt_authentication() throws Exception {
...
+        credentials.setUsername(JWT_USER.name());
+
+        Role authenticatedRole = jwtAuth.authenticate(credentials, null);
+        assertThat(authenticatedRole).isNotNull();
```
Comment (seut) [inline server/src/test/java/io/crate/auth/UserAuthenticationMethodTest.java:177]: The following assertions will fail if `authenticationRole` is null. No strong opinion, but I'd prefer to only assert for the minimum required for the test.
```suggestion
```

File: server/src/test/java/io/crate/auth/UserAuthenticationMethodTest.java [hunk 7]
```diff
@@ -145,25 +148,90 @@ public void test_jwt_authentication() throws Exception {
...
+                // User doesn't have "aud" JWT property, cluster id will be used as aud.
+                new JwtProperties("https://console.cratedb-dev.cloud/api/v2/meta/jwk/", "cloud_user", null)
+            )
+        );
+        JWTAuthenticationMethod jwtAuth = new JWTAuthenticationMethod(
+            roles,
+            jwkProviderFunction(null),
+            () -> clusterId
+        );
+        assertThat(jwtAuth.name()).isEqualTo("jwt");
```
Comment (seut) [inline server/src/test/java/io/crate/auth/UserAuthenticationMethodTest.java:200]:
```suggestion
```

File: server/src/test/java/io/crate/auth/UserAuthenticationMethodTest.java [hunk 8]
```diff
@@ -145,25 +148,90 @@ public void test_jwt_authentication() throws Exception {
...
+
+        Credentials credentials = new Credentials(JWT_TOKEN);
+        assertThat(credentials.username()).isNull();
```
Comment (seut) [inline server/src/test/java/io/crate/auth/UserAuthenticationMethodTest.java:203]:
```suggestion
```

**Comment by BaurzhanSakhariev** [comment]
Another unrelated failure
https://jenkins.crate.io/job/CrateDB/job/crate_on_pr/13453/testReport/junit/io.crate.integrationtests/LogicalReplicationITest/test_drop_and_recreate_subscription_while_published_table_is_written_to/

retest this please

File: server/src/test/java/io/crate/integrationtests/RoleManagementIntegrationTest.java [hunk 9]
```diff
@@ -283,30 +284,32 @@ public void testDropSuperUserThrowsException() {
...
         execute("CREATE USER user1 WITH (password = 'pwd', jwt = {\"iss\" = 'issuer1', \"username\" = 'user1'})");
         execute("CREATE USER user2 WITH (password = 'pwd', jwt = {\"iss\" = 'issuer2', \"username\" = 'user2'})");
 
+        //"aud" = 'aud11'
         // Regular alter - update all properties
-        execute("ALTER USER user1 set (jwt = {\"iss\" = 'issuer11', \"username\" = 'user11'})");
+        execute("ALTER USER user1 set (jwt = {\"iss\" = 'issuer11', \"username\" = 'user11', \"aud\" = 'aud11'})");
         execute("SELECT name, jwt from sys.users WHERE name = 'user1'");
-        assertThat(response).hasRows("user1| {iss=issuer11, username=user11}");
+        assertThat(response).hasRows("user1| {aud=aud11, iss=issuer11, username=user11}");
```
Comment (seut) [inline server/src/test/java/io/crate/integrationtests/RoleManagementIntegrationTest.java:307]: How are these statements related to the uniqueness validation? Can't we just leave them out?

**Comment by BaurzhanSakhariev** [comment]
we can skip `withAudience` in the verification builder if token doesn't have `aud` AND user doesn't have `aud`. Then it will be not required. Kept `aud` required in token in this version but if needed, it's a quick change.

File: server/src/main/java/io/crate/auth/HostBasedAuthentication.java (hunk 4 above)
Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/auth/HostBasedAuthentication.java:101]: removed

**Review by seut — APPROVED** [review]
👍
